include/tvm/relay/optimizer.h

```diff
@@ -0,0 +1,338 @@
+enum PassKind : int {
+  kModuleKind = 1,
+  kFunctionKind = 2
+};
+
+class PassContext;
+
+/*
+ * \brief PassContextNode contains the information that a pass can rely on, such as
+ * the analysis result.
+ */
+class PassContextNode : public RelayNode {
+ public:
+  /*!
+   * \brief The error reporter used to notify users why an optimization fails.
+   */
+  ErrorReporter err_reporter_;
+
+  PassContextNode() = default;
+
+  void VisitAttrs(tvm::AttrVisitor* v) final {
+  }
+
+  TVM_DLL static PassContext make();
+
+  static constexpr const char* _type_key = "relay.PassContext";
+  TVM_DECLARE_NODE_TYPE_INFO(PassContextNode, RelayNode);
+};
+
+class PassContext : public NodeRef {
+ public:
+  PassContext() = default;
+  explicit PassContext(NodePtr<tvm::Node> p) : NodeRef(p) {}
+
+  const PassContextNode* operator->() const {
+    return static_cast<PassContextNode*>(this->node_.get());
+  }
+
+  using ContainerType = PassContextNode;
+};
+
+// We use currying here. The Relay module is captured for optimizations. It
+// runs on a Relay node type NodeT and yields a new node with the same type.
+// For example, PassFunc<Function> indicates we perform a Function to Function
+// transformation on the given Module.
+template <typename NodeT,
+          typename = std::enable_if<(std::is_same<NodeT, Module>::value ||
+                                     std::is_same<NodeT, Function>::value)>>
+using PassFunc =
+    runtime::TypedPackedFunc<runtime::TypedPackedFunc<NodeT(NodeT)>(
+        const Module& mod)>;
+
+class Pass;
+
+/*!
+ * \brief PassNode is the base type of differnt types of optimization passes.
+ * It is implemented by different pass subclasses at different granularity of
+ * Relay nodes.
+ */
+class PassNode : public RelayNode {
+ public:
+  /*! \brief The name of an optimization/analysis pass. */
+  std::string name;
+  /*! \brief The minimal optimization level that this pass will be enabled. */
+  int opt_level;
+  /*! \brief The kind of an optimization/analysis pass. */
+  PassKind pass_kind;
+  /*! \brief The flag to indicate if a pass is enabled. */
+  bool enabled;
+  /*!
+   * \brief The passes that are required by this pass.
+   * TODO(zhiics) required_passes are used to identify the dependency of
+   * different passes. We will use it build the pass dependency graph in the
+   * followup PRs.
+   */
+  tvm::Array<tvm::Expr> required_passes;
+
+  /*!
+   * \brief Get the required passes for this pass as a vector of std::string.
+   */
+  TVM_DLL std::vector<std::string> RequiredPasses() const;
+
+  /*!
+   * \brief Execute the optimization pass using a functor. This functor invokes
+   * the `run` method to perform a real optimization on a certain type of node.
+   *
+   * \param mod The module that an optimization pass runs on.
+   * \param pass_ctx The context information that is used to help perform
+   *        a given pass.
+   */
+  void operator()(Module* mod, const PassContext& pass_ctx) {
```

Comment:
I wonder if we should make this a static method on the class?

Reply:
I think functors should be either non-static or free functions